accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,119 @@
+# Feature switch
+The functionality available in the .NET libraries is getting bigger and wider with every release. Applications typically don't need or want all of it, which means when an app includes the libraries as part of it, app size is unnecessarily increased. With technologies like the linker, we have the ability to remove parts of the framework which the app doesn't use. These technologies have limitations based on existing public API patterns and code behavior, sometimes they can't determine what the app really needs. This proposal adds a way to explicitly disable functionality in the framework.
+
+## Goals
+* The ability for developers to declaratively remove functionality in the framework, 3rd party libraries or final app
+* Consistent behavior across different build configurations - the app should get the exact same behavior whether the linker was used on it or not
+* Ability to set different defaults based on the type of application (Blazor, Xamarin, …)
+
+
+## Expected usage
+### Size
+By disabling functionality which is otherwise always available the linker can trim more code from the app reducing its size.
```

Comment:
What exactly will be a feature switch is yet to be determined. The one problem with having lot of them is testing - it makes the potential matrix of available configuration very large. So we'll probably try to keep the number small.

For the scenarios which you mention, ideally linker should be able to remove the code if you're not using it. Now making that a reality is pretty hard, but introducing a feature switch for those will be even harder.

Reply:
> For the scenarios which you mention, ideally linker should be able to remove the code if you're not using it. Now making that a reality is pretty hard, but introducing a feature switch for those will be even harder.

True that. Still, it will be useful in AoT builds. I would not worry about compat since its a one time build. You are not going to manually patch the native images.